src/logger/fileWriter.c

```diff
@@ -244,94 +260,152 @@ static int openNewLogfile(char *filename)
         pr_error_int_msg("File open err: ", rc);
-        LED_enable(3);
         return WRITING_INACTIVE;
     }
 
     return WRITING_ACTIVE;
 }
 
+struct file_status
+{
+        portTickType flush_tick;
+        portTickType write_tick;
+        bool logging;
+        enum writing_status writing_status;
+        char name[FILENAME_LEN];
+};
+
+static int logging_start(struct file_status *fs)
+{
+        if (fs->writing_status != WRITING_INACTIVE)
+                return 2;
+
+        pr_debug("Logging: Start\r\n");
+
+        fs->flush_tick = xTaskGetTickCount();
+        fs->write_tick = 0;
+        fs->writing_status = openNewLogfile(fs->name);
+        fs->logging = true;
+
+        return fs->writing_status == WRITING_ACTIVE ? 0 : 1;
+}
+
+static int logging_stop(struct file_status *fs)
+{
+        endLogfile();
+        fs->writing_status = WRITING_INACTIVE;
+        fs->logging = false;
+
+        logging_led_off();
+        return 0;
+}
+
+static bool remount_log_file(const struct file_status *fs)
+{
+        pr_error("Remounting FS due to write error.\r\n");
+        // XXX: Do we need to worry about rolling over to new files?
+
+        f_close(g_logfile);
+        UnmountFS();
+        InitFS();
+
+        return openLogfile(g_logfile, fs->name) == 0;
+}
+
+static int logging_sample(struct file_status *fs, LoggerMessage *msg)
+{
+        /* If we haven't starting logging yet, then no op */
+        if (!fs->logging)
+                return 0;
+
+        if (fs->writing_status != WRITING_ACTIVE)
+                return 1;
+
+        if (0 == fs->write_tick)
+                writeHeaders(msg->channelSamples, msg->sampleCount);
+
+        fs->write_tick++;
+        int attempts = 2;
+        int rc;
+        while (attempts--) {
+                rc = writeChannelSamples(msg->channelSamples,
+                                         msg->sampleCount);
+                if (rc == WRITE_SUCCESS)
+                        break;
+
+                if (attempts)
+                        remount_log_file(fs);
+        }
+
+        logging_led_toggle();
+        return rc == WRITE_SUCCESS ? 0 : 2;
+}
+
+static void flush_logfile(struct file_status *fs)
+{
+        if (fs->writing_status != WRITING_ACTIVE)
+                return;
+
+        if (!isTimeoutMs(fs->flush_tick, FLUSH_INTERVAL_MS))
+                return;
+
+        flushLogfile(g_logfile);
+        fs->flush_tick = xTaskGetTickCount();
+}
+
 void fileWriterTask(void *params)
 {
-    LoggerMessage *msg = NULL;
-    unsigned int flushTimeoutInterval = 0;
-    portTickType flushTimeoutStart = 0;
-    size_t tick = 0;
-    enum writing_status writingStatus = WRITING_INACTIVE;
-    char filename[FILENAME_LEN];
-
-    while(1) {
+        LoggerMessage *msg = NULL;
+        struct file_status fs = { 0 };
+
         while(1) {
-            //wait for the next sample record
-            xQueueReceive(g_sampleRecordQueue, &(msg), portMAX_DELAY);
+                int ok = 0;
```

Comment:
The variable `ok` used to be a `bool` until I had to debug an issue before I pushed this patch up for review.  I'll clean it up.